### PR TITLE
Refactor example token contracts

### LINKED ISSUE #109: Consolidate Token Create function contract examples
### Problem

In providing examples for HIP 206 smart contract both [TokenCreate](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/TokenCreate.sol) and [TokenCreateContract](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/examples/token-create/TokenCreateContract.sol)

It's not clear why some functions are in some files and not the others

### Solution

Consolidate the coverage of HIP 206 functions implemented in [IHederaTokenService](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol) into one single file [TokenCreateContract](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/examples/token-create/TokenCreateContract.sol).

The functions will include
- [ ] [createFungible](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L413)
- [ ] [createFungibleTokenWithCustomFees](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L428)
- [ ] [createNonFungible](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L440)
- [ ] [createNonFungibleTokenWithCustomFees](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L451)
- [ ] [mintToken](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L322)
- [ ] [associateTokens](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L366)
- [ ] [associateToken](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L373)
- [ ] [unfreezeToken](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L706)
- [ ] [grantKyc](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L714)
- [ ] [unpauseToken](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L734)

In this way `TokenCreateContract.sol` can contain all functions necessary for an admin/create to create tokens and allow users to transfer

### Alternatives

_No response_

### LINKED ISSUE #110: Add a TokenTransferContract example
### Problem

Currently trasnfer function examples in mixed in with token create and this results in a large contract and uncertainty around coverage

### Solution

Add a `contracts/hts-precompile/examples/token-transfer/TokenTransferContract.sol` example contract with coverage similar to `TokenCreateContract` and `TokenManageContract`

Functions to cover should include
- [ ] [cryptoTransfer](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L307) - ensure it supports sending HBAR, FT and NFT in one transaction
- [ ] [transferTokens](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L465)
- [ ] [transferNFTs](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L476)
- [ ] [transferToken](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L490)
- [ ] [transferNfT](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L504)
- [ ] [transferFrom](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L532) - Update `HederaTokenService.sol`
- [ ] [transferFromNfT](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L566)
- [ ] ~[approve](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L518)~ - should be under management
- [ ] ~[approveNFT](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L553)~ - should be under management
- [ ] ~[allowance](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L541)~ - should be under query
- [ ] ~[getApproved](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L574)~ - should be under query
- [ ] [setApprovalForAll](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L584)
- [ ] [isApprovedForAll](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L597)

Note: Some example coverage for transfer scenarios exists under [TokenCreate](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/TokenCreate.sol) and [TokenCreateContract](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/examples/token-create/TokenCreateContract.sol).
These may be moved over so no transfer coverage exist in the previous files.
References in other files may need to be update.

### Alternatives

_No response_

### LINKED ISSUE #111: Update TokenManagementContract
### Problem

[TokenManagment](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/examples/token-manage/TokenManagementContract.sol) does not contain complete coverage of management functions in HIP 514 and 206

### Solution

Ensure `TokenManagementContract.sol` has coverage for

- [ ] [deleteToken](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L624)
- [ ] [freezeToken](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L698)
- [ ] [revokeTokenKyc](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L722)
- [ ] [pauseToken](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L729)
- [ ] [wipeTokenAccount](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L741)
- [ ] [wipeTokenAccountNFT](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L752)
- [ ] [updateTokenInfo](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L762)
- [ ] [updateTokenExpiryInfo](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L770)
- [ ] [updateTokenKeys](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L778)
- [ ] [burnToken](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L343) 
- [ ] [dissociateTokens](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L395)
- [ ] [dissociateToken](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L402)
- [ ] [approve](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L518)
- [ ] [approveNFT](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L553) 

Note some have existing coverage in `TokenCreate.sol` or `TokenCreateContract.sol` and can be moved over from there.

In this way `TokenManagementContract.sol` can be seen as containing the functions necessary to modify a token or prevent users from transferring.

### Alternatives

_No response_

### LINKED ISSUE #112: Add a TokenInfoQuery Contract example
### Problem

Currently token query function via IHederaTokenService.sol are located in multiple files.
To reduce the size of the example contract and to allow for clean separation we should have a separte contract for just queries

### Solution

Add a `contracts/hts-precompile/examples/token-query/TokenQueryContract.sol` with coverage for

- [ ] [allowance](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L541)
- [ ] [getApproved](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L574)
- [ ] [isApprovedForAll](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L597)
- [ ] [isFrozen](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L608)
- [ ] [isKyc](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L617)
- [ ] [getTokenCustomFees](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L632)
- [ ] [getTokenDefaultFreezeStatus](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L640)
- [ ] [getTokenDefaultKycStatus](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L648)
- [ ] [getTokenExpiryInfo](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L656)
- [ ] [getFungibleTokenInfo](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L664)
- [ ] [getTokenInfo](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L672)
- [ ] [getTokenKey](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L681)
- [ ] [getNonFungibleTokenInfo](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L690)
- [ ] [isToken](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L786)
- [ ] [getTokenType](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L794)
- [ ] [allowance](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L541) 
- [ ] [getApproved](https://github.com/hashgraph/hedera-smart-contracts/blob/main/contracts/hts-precompile/IHederaTokenService.sol#L574)

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by georgi-l95**
**Description**:
Refactor example contracts, so that each contract is responsible for one type of operations.
Apply those changes in the tests.

**Related issue(s)**:

Fixes #109  
Fixes #110 
Fixes #111  
Fixes #112  

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

File: contracts/hts-precompile/examples/token-create/TokenCreateContract.sol [hunk 1]
```diff
@@ -140,61 +143,88 @@ contract TokenCreateContract is TokenCreate {
...
+
+
+
+
+
+
+
+
+
+
```
Comment (dimitrovmaksim) [inline contracts/hts-precompile/examples/token-create/TokenCreateContract.sol:229]: Lots of empty lines :D

Comment (georgi-l95) [reply contracts/hts-precompile/examples/token-create/TokenCreateContract.sol:229]: Thank you. fixed

**Comment by dimitrovmaksim** [comment]
Would any contracts here be affected by the changes in this PR https://github.com/hashgraph/hedera-smart-contracts/pull/99 and the resolving of this issue https://github.com/hashgraph/hedera-smart-contracts/issues/85?

File: contracts/hts-precompile/examples/token-transfer/TokenTransferContract.sol [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    function transferFromNFTPublic(address token, address from, address to, uint256 serialNumber) public returns (int64 responseCode) {
+        responseCode = this.transferFromNFT(token, from, to, serialNumber);
+        emit ResponseCode(responseCode);
+
+        if (responseCode != HederaResponseCodes.SUCCESS) {
+            revert ();
+        }
+    }
+
+    function approvePublic(address token, address spender, uint256 amount) public returns (int responseCode) {
```
Comment (Nana-EC) [inline contracts/hts-precompile/examples/token-transfer/TokenTransferContract.sol:77]: Approves should be under `TokenManagement` I may have missed this initially.
The operation doesn't transfer

**Review by Nana-EC — CHANGES_REQUESTED** [review]
Looking good, appreciate the quick turn around.
Some items highlighted in comments

File: contracts/hts-precompile/examples/token-transfer/TokenTransferContract.sol (hunk 2 above)
Comment (georgi-l95) [reply contracts/hts-precompile/examples/token-transfer/TokenTransferContract.sol:77]: I've moved `isApprovedForAllPublic`, `getApprovedPublic` to query contract and `approvePublic`, `approveNFTPublic` and `setApprovalForAllPublic` to managment.

File: contracts/hts-precompile/examples/token-transfer/TokenTransferContract.sol [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    function approveNFTPublic(address token, address approved, uint256 serialNumber) public returns (int responseCode) {
+        responseCode = HederaTokenService.approveNFT(token, approved, serialNumber);
+        emit ResponseCode(responseCode);
+
+        if (responseCode != HederaResponseCodes.SUCCESS) {
+            revert ();
+        }
+    }
+
+    function allowancePublic(address token, address owner, address spender) public returns (int responseCode, uint256 amount) {
```
[the remark this answers is not in the document]
Comment (georgi-l95) [reply contracts/hts-precompile/examples/token-transfer/TokenTransferContract.sol:95]: Yes, that's right, I've removed it from transfer contract, it was already in query contract.

File: contracts/hts-precompile/HederaTokenService.sol [hunk 4]
```diff
@@ -271,6 +271,37 @@ abstract contract HederaTokenService is HederaResponseCodes {
...
+    /// Only applicable to NFT tokens
+    /// @param token The address of the non-fungible Hedera token to transfer
+    /// @param from The account address of the owner of `serialNumber` of `token`
+    /// @param to The account address of the receiver of `serialNumber`
+    /// @param serialNumber The NFT serial number to transfer
+    /// @return responseCode The response code for the status of the request. SUCCESS is 22.
+    function transferFromNFT(address token, address from, address to, uint256 serialNumber) external returns (int64 responseCode)
+    {
+        (bool success, bytes memory result) = precompileAddress.call(
+            abi.encodeWithSelector(IHederaTokenService.transferFrom.selector,
```
[the remark this answers is not in the document]
Comment (georgi-l95) [reply contracts/hts-precompile/HederaTokenService.sol:300]: Fixed. Thank you.

**Review by Nana-EC — APPROVED** [review]
LG

**Pull request merged**